### ISSUE TITLE
Help Required for enabling the default option for Owner in The Repository Picker using Software Templates.

### ISSUE THREAD

**Issue by aswathysen** [backstage/backstage#8047]
#helpwanted
A solution required to have an option to make Owner in The Repository Picker as default instead of manually adding the Owner. 
This is regarding the Software template: https://backstage.io/docs/features/software-templates/writing-templates#the-repository-picker

**Comment by freben** [comment]
Hi! Could you explain the use case a bit more? Who / what should the value be, is there some particular logic? This might be solvable with a custom picker component either way, and that could be a good contribution.

**Comment by aswathysen** [comment]
@freben , Hi!

The Use case is as follows:
 **Existing logic for repository location**

`- title: Choose a location
  required:
    - repoUrl
  properties:
    repoUrl:
      title: Repository Location
      type: string
      ui:field: RepoUrlPicker
      ui:options:
        allowedHosts:
          - github.com`
RepoUrlPicker will return an object so parameters.repoUrl  as { host: string; owner: string; repo: string }

`Output format : { host: string; owner: string; repo: string }`

From this output we are able to get the owner and repo details the user has passed. 

**Requirement**
If the user is belonging to an organization the owner remains the same, is there a possibility to default the owner (without hardcoding the owner in the url directly) , rather than making the user to fill the owner details each and every time they use the software template.

**An example :**

`- title: Choose a location
  required:
    - repoUrl
  properties:
    repoUrl:
      title: Repository Location
      type: string
      ui:field: RepoUrlPicker
      ui:options:
        allowedHosts:
          - github.com 
        allowedOwner(or Owners):
          - ABC`

**Comment by benjdlambert** [comment]
#8207 is the PR I believe for this ticket.

**Comment by OrKarstoft** [comment]
@benjdlambert I believe what's being asked for is a way to default the owner field.
E.g. If I'm in group "Finance" on Github and I go to and create a new service. The owner field should default to "Finance" so the users don't have to retype that every time.

I spot one issue though - what if you're in multiple groups?
Maybe just pick the first.

**Comment by benjdlambert** [comment]
So maybe we can leave this open for now to track the work, but I believe the solution here is to use the users oauth token when they are signed in with github etc, to grab the teams or owners then belong to right and automatically add it to the `RepoPicker` owner part right?

**Comment by aswathysen** [comment]
This

> So maybe we can leave this open for now to track the work, but I believe the solution here is to use the users oauth token when they are signed in with github etc, to grab the teams or owners then belong to right and automatically add it to the `RepoPicker` owner part right?

This solution seems to be more dynamic in structure , then on usage of the software template , it is automatically picked up with their sign-in oauth token!

**Comment by benjdlambert** [comment]
Yep - so we've had this in mind for a while, #8328 is starting the work off that shows how to grab a users token. So hopefully we can build on that and add this feature after. I'd also really like to break apart and refactor the `RepoUrlPicker` component as it's just growing in complexity and becoming harder to reason about what its going on there too.

**Comment by jrusso1020** [comment]
We created a custom scaffolder extension for this, but it's a little customized to how we represent our groups. The gist is the same though we have a selector/dropdown that grabs all groups, then we grab the current user's identity and use their group as the default option when selecting an owner

```
import React, { useCallback, useEffect } from 'react';
import { identityApiRef, useApi } from '@backstage/core-plugin-api';
import {
  catalogApiRef,
  humanizeEntityRef,
} from '@backstage/plugin-catalog-react';
import { parseEntityRef } from '@backstage/catalog-model';
import useAsync from 'react-use/lib/useAsync';
import { FieldProps } from '@rjsf/core';
import {
  createScaffolderFieldExtension,
  scaffolderPlugin,
} from '@backstage/plugin-scaffolder';
import { TextField } from '@material-ui/core';
import FormControl from '@material-ui/core/FormControl';
import Autocomplete from '@material-ui/lab/Autocomplete';

export const DefaultOwnerPicker = (props: FieldProps<string>) => {
  const {
    onChange,
    schema: { title = 'Owner', description = 'The owner of the component' },
    required,
    rawErrors,
    formData,
    idSchema,
  } = props;
  const allowedKinds = ['Group'];
  const defaultKind = 'Group';

  const catalogApi = useApi(catalogApiRef);

  const { value: entities, loading } = useAsync(() =>
    catalogApi.getEntities(
      allowedKinds ? { filter: { kind: allowedKinds } } : undefined,
    ),
  );

  const entityRefs = entities?.items.map(e =>
    humanizeEntityRef(e, { defaultKind }),
  );

  const onSelect = useCallback(
    (_: any, value: string | null) => {
      onChange(value || '');
    },
    [onChange],
  );

  useEffect(() => {
    if (entityRefs?.length === 1) {
      onChange(entityRefs[0]);
    }
  }, [entityRefs, onChange]);

  // get the initial default value for the team
  const identityApi = useApi(identityApiRef);
  useAsync(async () => {
    const identity = await identityApi.getBackstageIdentity();
   
   // CUSTOM LOGIC THAT GRABS THEIR TEAM TO SET AS DEFAULT
    const userTeam = identity.ownershipEntityRefs.find(ref => {
      const entityRef = parseEntityRef(ref);
      return entityRef?.name.startsWith('team-');
    });

    if (userTeam === undefined) {
      return;
    }

    const entityRef = parseEntityRef(userTeam);

    onChange(entityRef?.name);
  });

  return (
    <FormControl
      margin="normal"
      required={required}
      error={rawErrors?.length > 0 && !formData}
    >
      <Autocomplete
        disabled={entityRefs?.length === 1}
        id={idSchema?.$id}
        value={(formData as string) || ''}
        loading={loading}
        onChange={onSelect}
        options={entityRefs || []}
        autoSelect
        freeSolo={false}
        renderInput={params => (
          <TextField
            {...params}
            label={title}
            margin="normal"
            helperText={description}
            variant="outlined"
            required={required}
            InputProps={params.InputProps}
          />
        )}
      />
    </FormControl>
  );
};

export const DefaultOwnerPickerExtension = scaffolderPlugin.provide(
  createScaffolderFieldExtension({
    name: 'DefaultOwnerPicker',
    component: DefaultOwnerPicker,
  }),
);
```

**Comment by benjdlambert** [comment]
Related https://github.com/backstage/backstage/pull/11551

**Comment by benjdlambert** [comment]
Maybe one for after #10430

**Comment by benjdlambert** [comment]
Going to use this ticket as a base here for a little more features.

What I think we want to do here is have an `@backstage/integrations-backend-plugin` that can do a few things, and that plugin will be called by the `RepoUrlPicker` and other `customFieldExtensions` to grab the owners and organizations that are available to the user.

If no token is provided from the frontend, then it will use the `integrationsConfig` for the specified host, (this might include GitHub apps for instance too), or if a token is provided then it will go an grab the owners and orgs that are available to the token for creation.

**Comment by RSG-12** [comment]
We are trying the similar implementation using the custom field extensions. 

We have 2 custom fields in the template, once user enters some value in the first field we invoke scmAuthApi and pop up the Gitlab auth. After successful auth by user we are trying to grab the token and invoke an api to fetch subgroups belonging to user and load them in another custom field (a dropdown). However, we are not able to call the load the dropdown field with api response. Any suggestions on this one or you have a better approach, please let us know.

**Comment by benjdlambert** [comment]
@RSG-12 i'm not sure I full understand what the issue is, can you put together an example reproduction or some sample code that explains the problem and what you expect to happen? :pray:
